src/metabase/sync_database/analyze.clj

```diff
@@ -7,236 +7,154 @@
-
-;; TODO - It's weird that this one function requires other functions as args when the whole rest of the Metabase driver system
-;;        is built around protocols and record types. These functions should be put back in the `IDriver` protocol (where they
-;;        were originally) or in a special `IAnalyzeTable` protocol).
-(defn make-analyze-table
-  "Make a generic implementation of `analyze-table`."
-  {:style/indent 1}
-  [driver & {:keys [field-avg-length-fn field-percent-urls-fn calculate-row-count?]
-             :or   {field-avg-length-fn   (partial driver/default-field-avg-length driver)
-                    field-percent-urls-fn (partial driver/default-field-percent-urls driver)
-                    calculate-row-count?  true}}]
-  (fn [driver table new-field-ids]
-    (let [driver (assoc driver :field-avg-length field-avg-length-fn, :field-percent-urls field-percent-urls-fn)]
-      {:row_count (when calculate-row-count? (u/try-apply table-row-count table))
-       :fields    (for [{:keys [id] :as field} (table/fields table)]
-                    (let [new-field? (contains? new-field-ids id)]
-                      (cond->> {:id id}
-                               (test-for-cardinality? field new-field?) (test:cardinality-and-extract-field-values field)
-                               new-field?                               (test:new-field driver field))))})))
-
-(defn generic-analyze-table
-  "An implementation of `analyze-table` using the defaults (`default-field-avg-length` and `field-percent-urls`)."
-  [driver table new-field-ids]
-  ((make-analyze-table driver) driver table new-field-ids))
-
-
+(defn- percent-match
+  [predicate values]
+  (if-let [non-nil-values (remove nil? values)]
+    (if (seq non-nil-values)
+      (as-> non-nil-values x
+        (filter predicate x)
+        (count x)
+        (/ x (count non-nil-values))
+        (* x 100)
+        (int x))
+      0)
+    0))
+
+(def ^:private percent-json  (partial percent-match value-is-valid-json?))
+(def ^:private percent-email (partial percent-match value-is-valid-email?))
+
+(defn- percent-valid-urls
+  "Recursively count the values of non-nil values in VS that are valid URLs, and return it as a percentage."
+  [vs]
+  (loop [valid-count 0, non-nil-count 0, [v & more :as vs] vs]
+    (cond (not (seq vs)) (if (zero? non-nil-count) 0.0
+                             (float (/ valid-count non-nil-count)))
+          (nil? v)       (recur valid-count non-nil-count more)
+          :else          (let [valid? (and (string? v)
+                                           (u/is-url? v))]
+                           (recur (if valid? (inc valid-count) valid-count)
+                                  (inc non-nil-count)
+                                  more)))))
+
+(defn field-avg-length
+  "Default implementation of optional driver fn `field-avg-length` that calculates the average length in Clojure-land via `field-values-lazy-seq`."
```

Comment:
...functions whose purpose has completely changed but without corresponding updates to the docstrings. Makes it hard to figure out how this is all supposed to work 😓